Step5 (JEJE)

jung-yun opened the pull request:
Step5를 구현했습니다.

예전에 UISegmentedControl 구현 시 tintColor만 설정해주면 예시처럼 divider와 boarder에 색을 입힐 수 있었는데 iOS 버전업이 되면서 구현 방법이 바뀐건지 궁금합니다. 인터넷에서 예전 코드를 따라해봤을 때 예상하는 결과와 비슷하게 구현되지 않아서 찾아보니 이와 같은 정보를 봐서요. 아니면 구현 방법을 제가 놓치고 있는 건지 힌트 부탁드립니다 

횡방향 스택뷰를 종방향 스택뷰에 넣으면서 autolayout잡는데 아직 많이 미숙하다는 생각을 하게된 미션이었습니다. 여러가지 문제가 많지만 현재 구현에서 시뮬레이터를 똑바로 세우면 목표 화면에 가깝게 구현되는데 눕히면 레이아웃이 완전 깨지는 현상이 발견되고 있습니다.

ViewController에서 UISegmentedControl로 변수를 입력받다보니 일부 로직을 여기서 구현하게 돼서 객체지향지향적인 코딩에서 약간 벗어나가 있는 것 같아 아쉽습니다. 남는 시간에 개선이 필요한 것 같습니다.

jung-yun committed:
Fix access level

jung-yun committed:
Edit README

jung-yun committed:
Refacto code

Declare enum that contains all the possible inputs for withPlayerOf

jung-yun committed:
Create directories in accordance with MVC design pattern

jung-yun committed:
Add card image files

jung-yun committed:
Implement methods to dislay dealer's cards on screen

Added giveCardInfo method to pass cardsInHand in dealer class
Revised play method (previously playGame method) to have cardDashboard proeprty and pass it over to view controller
Revised makeCardImageView to generate card image name using card info

Lots of things to be improved, since these whole amendment were swiftly implemented.
I will more focus on OOP design pattern and clean code from next commit

jung-yun committed:
Refactor Code

Moved card info to card image conversion into Card class

jung-yun committed:
Implement data flow passing players' card info to viewController

jung-yun committed:
Remove obsolte parmeter on method

jung-yun committed:
Implement methods that show cards images on screen

Deleted unnecessary mothod
Re-set some constraints to rightfully display cards

jung-yun committed:
Create method to make card stack view including prize place

jung-yun committed:
Move makeCardImageView into Card class

jung-yun committed:
Remove obsolete test method

jung-yun committed:
Implement individual stack view

jung-yun committed:
Adjust size of prizeImageView and image

jung-yun committed:
Implement cardSelectSegmentedControl

jung-yun committed:
Implement numberOfPlayersSegmentedControl

jung-yun committed:
Test stackView autoLayout

jung-yun committed:
Implement dashboard stack view

jung-yun committed:
Add target to cardSelectSegmentedControl

jung-yun committed:
Add methods to reset player and dealer's card deck

jung-yun committed:
Refactor method name

jung-yun committed:
Add target to the second segmentedControl

jung-yun committed:
Refactor

jung-yun committed:
Implement game variance selection

Moved enums to view controller to make use of it(Any way to keep enums in PockerGame class?)
Changed initializer of PokerGame in accordance with enums that moved

jung-yun committed:
Adjust UI details and clean up code

jung-yun commented on PokerGameApp/PokerGameApp.xcodeproj/project.pbxproj:
```diff
@@ -51,6 +51,28 @@
 /* End PBXFrameworksBuildPhase section */
 
 /* Begin PBXGroup section */
+		88EEB4C325DE6A9700D1D840 /* Model */ = {
```
swift 파일들을 Model, View, Controller 디렉토리에 넣었습니다

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,214 @@
+import UIKit
+
+class ViewController: UIViewController {
+    enum NumberOfPlayers: Int {
+        case one = 1
+        case two
+        case three
+        case four
+    }
+    
+    enum StudVariant: Int {
+        case fiveCardStud = 5
+        case sevenCardStud = 7
+    }
```
NumberOfPlayers 나 StudVariant 는 ViewController보다 Game과 관련이 높은 타입인 것 같아요. 그러니까 Game과 같은 파일에 포함하는 게 좋겠네요.

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,214 @@
+    
+    
+    lazy var studSelectSegmentedControl: UISegmentedControl = {
+        let control = UISegmentedControl(items: ["7 cards", "5 cards"])
```
`["7 cards", "5 cards"]`는 StudVariant과 관련이 높은 값 같네요. 
Rank나 Suit 처럼 CaseIterable 프로토콜을 채택하면 하드코딩하지 않고 배열을 구할 수 있지 않을까요?

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,214 @@
+     }()
+    
+    lazy var numberOfPlayersSegmentedControl: UISegmentedControl = {
+        let control = UISegmentedControl(items: ["2명", "3명", "4명"])
```
`["2명", "3명", "4명"]` 도 마찬가지로 playerNumbers와 관련이 높아보입니다. 하드코딩하지 않고 playerNumbers에서 생성할 수 있을까요?

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        configureVerticalStackView()
+        guard let gameResult = game.play() else { return }
+        print(gameResult)
+        generateGameDashBoard(with: gameResult)
```
gameResult를 가져와서 전달하는 구조도 나쁘지 않습니다만. 값을 가져오지 않고 game에 generateGameDashBoard 내용을 클로저로 전달하는 방법도 있습니다. 그러면 카드 배열을 가져오지 않고, card만으로 뷰를 생성할 수 있습니다. 이 방법을 도전해보세요.

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -0,0 +1,214 @@
+        case 1:
+            studVariant = .fiveCardStud
+        default:
+            print("meh")
```
여기 switch-case 대신에 studVariant = StudVariant.init() 형태로 표현해보세요. 
default에 print 구문은 전혀 의미가 없습니다. 차라리 그냥 비워두세요. ;만 찍어도 됩니다. (switch구문을 빼면 안해도...)

godrm commented on PokerGameApp/PokerGameApp/Model/Card.swift:
```diff
@@ -0,0 +1,93 @@
+
+    public let suit: Suit, value: Value
+    
+    public func createCardImage() -> UIImage? {
+        var cardImageName: String {
+            var result: String = ""
+            switch self.suit {
+            case .clovers: result.append("c")
+            case .hearts: result.append("h")
+            case .spades: result.append("s")
+            case .diamonds: result.append("d")
+            }
+            switch self.value {
+            case .two: result.append("2")
+            case .three: result.append("3")
+            case .four: result.append("4")
+            case .five: result.append("5")
+            case .six: result.append("6")
+            case .seven: result.append("7")
+            case .eight: result.append("8")
+            case .nine: result.append("9")
+            case .ten: result.append("10")
+            case .ace: result.append("A")
+            case .jack: result.append("J")
+            case .queen: result.append("Q")
+            case .king: result.append("K")
+            }
+            return result
+        }
+        return UIImage(named: cardImageName) ?? nil
```
가장 하위에 있는 Card 가 UIKit 을 포함하는 것은 좋은 선택이 아닙니다.
UIKit은 출력을 위한 뷰 프레임워크라서 이런 하위 모델과는 분리하는 게 좋습니다. 
이렇게 파일이름을 처리하는 것은 상위 모듈 중에서 처리해야 합니다. 다른 곳으로 옮겨보세요.

godrm commented on PokerGameApp/PokerGameApp/Model/Card.swift:
```diff
@@ -0,0 +1,93 @@
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        imageView.contentMode = .scaleAspectFit
+        return imageView
+    }
```
여기 makeCardImageView 도 마찬가지로 뷰와 관련된 코드라서 분리하세요.

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -16,6 +16,14 @@ class Dealer: Receiver {
+    
+    func cardInfo() -> Array<Card> {
+        return cardsInHand
+    }
```
위에 언급한 것처럼 배열을 전달하지 말고, 매개변수로 클로저를 전달하는 구조로 변경해보세요.

godrm reviewed (changes requested):
주말인데도 수고하셨습니다. 다음 피드백을 확인해보시고, 개선해보세요.

jung-yun committed:
Update README

jung-yun committed:
Refactor Code

Moved StudVariant and NumberOfPlayers Enums into PokerGame class since they are more related with it
Applied CaseInterable to both enums
Created export-Enums-Options methods and applied CustomStringConvertible to enums to apply enum values to ViewController UISegmented Controls

godrm commented on PokerGameApp/PokerGameApp/Controller/ViewController.swift:
```diff
@@ -174,29 +162,13 @@ class ViewController: UIViewController {
-        default:
-            print("meh")
-        }
+        game.selectStudVariant(index: sender.selectedSegmentIndex)
```
아.. Int 타입말고 StudVariant.init(), NumberOfPlayers.init() 을 해서 세부 타입을 넘겼으면 좋았을 것 같네요.

jung-yun replied:
아하 그렇군요.... 😵 수정하도록 하겠습니다

godrm commented:
주말인데 수정하느냐고 수고하셨습니다.

jung-yun commented:
JK야말로 주말인데 리뷰 봐주신다고 고생이 많으십니다🙏  아직 수정 완료가 아니라 아쉽네요 ㅠ... 앞으로도 틈틈히 리뷰 해주신거 다 반영하도록 하겠습니다. 감사합니다~

godrm commented:
네. 이제 다음주부터는 저도 천천히 리뷰하려구요 ㅎㅎ 다들 열심히 하길래... 내일 봐요
